topics/go/algorithms/data/heap/heap.go

```diff
@@ -0,0 +1,140 @@
+	Value int
+	Index int
+}
+
+// Heap represents a list of data.
+type Heap struct {
+	size     int
+	capacity int
+	dataHeap *dataHeap
+}
+
+// New returns a new Heap with given capacity in which we will store our data.
+func New(cap int) (*Heap, error) {
+
+	if cap <= 0 {
+		return nil, errors.New("invalid capacity")
+	}
+
+	h := Heap{
+		size:     0,
+		capacity: cap,
+		dataHeap: &dataHeap{},
+	}
+
+	return &h, nil
+}
+
+// Store is storing the element of type Data in the heap.
+func (h *Heap) Store(data *Data) error {
+
+	// If we try to add data to full Heap, we will return an error.
+	if h.size >= h.capacity {
+		return errors.New("getting out of heap capacity")
+	}
+
+	// The complexity of the heap.Push operation is O(log n) where n = h.dataHeap.Len().
+	// The element which we will be store in the heap will be added closely to the end of the heap and after what it
+	// will be sift up, until it not get in the right position in the heap according it's value.
+	heap.Push(h.dataHeap, data)
+	h.size++
+	return nil
+}
+
+// Extract returns the minimum element of type Data from the heap or error error heap is empty.
+// Since we have minimum heap implementation. The size of the heap will be decreased.
+func (h *Heap) Extract() (*Data, error) {
+
+	// Check that heap is not empty
+	if len(*h.dataHeap) == 0 {
+		return nil, ErrEmptyHeap
+	}
+
+	// heap.Pop return the element from the heap. the type of element is empty interface.
+	element := heap.Pop(h.dataHeap)
+	h.size--
+
+	// Since the returned value we got has an empty interface type, let's convert it to our Data type.
+	data := element.(*Data)
+
+	return data, nil
+}
+
+// Remove removes given element of type Data from the heap. The size value of the Heap will be decreased.
+// Returns error when heap is empty.
+func (h *Heap) Remove(data *Data) error {
+
+	// Check that heap is not empty
+	if len(*h.dataHeap) == 0 {
+		return ErrEmptyHeap
+	}
+
+	// Remove given data from the heap
+	heap.Remove(h.dataHeap, data.Index)
+	h.size--
+
+	return nil
+}
+
+// GetRoot returns the element which is stored in the root of the heap and does not change the heap.
+// Returns error if heap is empty.
+func (h *Heap) GetRoot() (data *Data, err error) {
+
+	// Check that heap is not empty
+	if len(*h.dataHeap) == 0 {
+		return nil, ErrEmptyHeap
+	}
+
+	return (*h.dataHeap)[0], nil
+}
+
+// Size return the current size of the Heap
+func (h *Heap) Size() int {
+	return h.size
+}
+
+// dataHeap is a type which implements a minimum heap interface from the container/heap package which means that data
+// with minimum values will be stored at the top of the heap.
+type dataHeap []*Data
+
+func (dh *dataHeap) Len() int {
+	return len(*dh)
+}
+
+func (dh *dataHeap) Less(i, j int) bool {
+	return (*dh)[i].Value < (*dh)[j].Value
+}
+
+func (dh *dataHeap) Swap(i, j int) {
+	(*dh)[i], (*dh)[j] = (*dh)[j], (*dh)[i]
+	(*dh)[i].Index = j
+	(*dh)[j].Index = i
+}
+
+func (dh *dataHeap) Push(x interface{}) {
```

Comment:
Let's use `v` for value instead of `x`

Reply:
This bothers me. This example defines the data and you are using that in he declaration of the type. Then you wrote these functions to use the empty interface, then you check we are using the right type. 

To keep it simple, these functions should ask for a Data value.

Since it is a heap, the idea is that this function provides the memory, not the other way around. 

If the user is providing the memory, then this is nothing more than a list of data items. If we can provide the memory for a Data value, that would be so much better. Then we can maintain the contiguous block of memory I want. Return the address to that.

Reply:
I like your proposal to use the contiguous block of memory, but in that case as I correctly understand we should refuse the usage of the heap realization from the standard library.

Let me try to explain what I mean, and with what I faced when trying to use only Data value and also a value semantics here.

When I change the implementation to use value semantic, every time I use heap.Push/heap.Pop method the program operates with a copy of the data heap slice. According to debug and tests, which are really helpful in that case, every time I get the empty slice. The value which we try to store in our heap is not storing there and dataHeap slice always has zero length and zero capacity. That is why I've not changed the implementation to use Data
and dataHeap value semantics.

But again I'm really like your proposal to use a contiguous block of memory there, so from my side, I can offer the following:

1) maybe we can change implementation and make our Heap type implements heap interface from the standard library. In that case we would be able to use a slice of Data instead use our custom dataHeap type.

2) we can implement our heap implementation on the slice without using a heap interface from the standard library.

3) we can just store the type primitives in our heap like int, string, float and in that case we will not need to create our dataHeap type of [] * Data